contracts/account/paymaster/PaymasterERC20.sol

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {ERC4337Utils, PackedUserOperation} from "@openzeppelin/contracts/account/utils/draft-ERC4337Utils.sol";
+import {IERC20, SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {EIP712} from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {PaymasterCore} from "./PaymasterCore.sol";
+import {AbstractSigner} from "../../utils/cryptography/AbstractSigner.sol";
+
+/**
+ * @dev Extension of {PaymasterCore} that enables users to pay gas with ERC-20 tokens.
+ */
+abstract contract PaymasterERC20 is PaymasterCore {
+    using ERC4337Utils for *;
+    using Math for *;
+    using SafeERC20 for IERC20;
+
+    event UserOperationSponsored(
+        bytes32 indexed userOpHash,
+        address indexed user,
+        address indexed guarantor,
+        uint256 tokenAmount,
+        uint256 tokenPrice,
+        bool paidByGuarantor
+    );
+
+    // Over-estimations: ERC-20 balances/allowances may be cold and contracts may not be optimized
+    uint256 private constant POST_OP_COST = 30_000;
+    uint256 private constant POST_OP_COST_WITH_GUARANTOR = 50_000;
```

Comment:
I think it's worth specifying how these are calculated. The values seems fine but if I'm a developer using the library, the first thing I'd ask myself is what happens if the underlying ERC-20 consumes even more gas than this over-estimation.

Reply:
It was estimated by "looking" at gas cost, and its relevance is "checked" in the unit tests:

```javascript
expect(tokenAmount)
  .to.be.greaterThan(actualGasCost * 2n)
  .to.be.lessThan((actualGasCost * 2n * 110n) / 100n);
```

If the underlying ERC20 conumes even more gas, then the result is that the cost of the PostOp, which is beared (in Eth) by the paymaster would not be fully refunded (in ERC20, by the account, to the paymaster). So the paymaster may lose a bit of money (or make less than expected). 

That is if the paymaster is accepting this token as payment (its the paymaster's decision to decide what ERC20 they accept)